### DISCUSSION TITLE
Possible to use only openssh client with TLS Routing enabled?

### DISCUSSION THREAD

**Discussion by cyberbliss** [gravitational/teleport#14005] [Q&A]
Hi

I've successfully set up my v9.3.7 Teleport service with TLS Routing enabled, meaning all traffic goes via port 443 on the server. I can ssh onto the SSH nodes using either tsh or ssh (when using a config generated by tsh config). However, I can't figure out how to do the same using only the openssh client, i.e. using a proxy command that does not call tsh. Is this even possible?

Thanks... Steve

**Comment by webvictim** [comment]
Yep, it's theoretically possible with the use of `openssl s_client` with `-alpn teleport-proxy-ssh` as a `ProxyCommand`:

```
#
# Begin generated Teleport configuration for teleport.example.com:443 from `tsh config`
#

# Common flags for all example hosts
Host *.example teleport.example.com
    UserKnownHostsFile "~/.ssh/teleport_known_hosts"
    #IdentityFile "~/.tsh/keys/teleport.example.com/admin"
    #CertificateFile "~/.tsh/keys/teleport.example.com/admin-ssh/example-cert.pub"

# Flags for the proxy only
Host teleport.example.com
    Port 3023
    ProxyCommand openssl s_client -quiet -connect teleport.example.com:443 -servername teleport.example.com -alpn teleport-proxy-ssh 2>/dev/null

# Flags for all other hosts except the proxy
Host *.example !teleport.example.com
    Port 3022
    #ProxyCommand "/usr/local/bin/tsh" proxy ssh --cluster=example --proxy=teleport.example.com %r@%h:%p
    ProxyJump %r@teleport.example.com
    #ProxyCommand ssh %r@teleport.example.com -s proxy:%h:%p

# End generated Teleport configuration
```

(h/t @programmerq for the `openssl` magic)

**Reply by cyberbliss** [reply]
Thanks for your speedy reply. I tried your suggestion and it successfully authenticates to the proxy but fails to connect to the actual SSH node. Looking in the logs on the Teleport server I see these messages:
```
User Message: no tunnel connection found: no node reverse tunnel for  found] trace.fields:map[cluster:teleport.example.com] reversetunnel/localsite.go:313
Jun 30 17:10:28 teleport-server teleport[28104]: 2022-06-30T17:10:28Z DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:276
Jun 30 17:10:58 teleport-server teleport[28104]: 2022-06-30T17:10:58Z DEBU [PROXY:SER] Error occurred while dialing directly. address:dev-server.teleport.example.com:3022 error:[
Jun 30 17:10:58 teleport-server teleport[28104]: ERROR REPORT:
Jun 30 17:10:58 teleport-server teleport[28104]: Original Error: *net.OpError dial tcp 150.230.118.191:3022: i/o timeout
```
`dev-server.teleport.example.com` is the SSH node and has no ports open to the Internet
`150.230.118.191` is the IP address of my teleport.example.com server and only port 443 is open to the Internet (so I'm not surprised that its getting an i/o timeout)

**Reply by webvictim** [reply]
Hmm, that's interesting. The issue is that normally when `tsh proxy ssh` runs, it's actually requesting the `proxy` subsystem from the Teleport server which will handle tunnelling through to the server itself (even when it isn't listening on port 3022, as it is the case when your nodes are joined to the proxy via reverse tunnel)

You'd likely need to find a way of providing the actual open channel from `openssl s_client` through to `ssh -s proxy:<hostname>:3022`. ~~I can't think of a way of doing that currently.~~

**Reply by webvictim** [reply]
Actually - I suspect the problem is that your SSH config isn't matching against `dev-server.teleport.example.com` because it says `*.example` in the text. Try changing that to `*.teleport.example.com` instead...

**Reply by cyberbliss** [reply]
Thanks for your suggestions. As it happens I had already changed `Host *.example...` to `Host *.teleport.example.com...` (I just forgot to mention that change).

**Reply by webvictim** [reply]
Cool, I was just making sure.

**Comment by webvictim** [comment]
Can you share the output of `ssh -vv dev-server.teleport.example.com` when it isn’t working?

**Comment by webvictim** [comment]
FYI there was a bug fix PR merged in Teleport 9.3.9 which might be relevant here: https://github.com/gravitational/teleport/pull/13928
